Fe start failed with RollbackException: must rollback xxx total commits(xxx of which were durable) to the earliest point

<!-- (At least include the following, feel free to add if you have more content) -->

### Steps to reproduce the behavior (Required)
Actually it is the transferring master that causes the result in the production environment.
But the cases that transfer master and remains log that not synchronized to other nodes is not easy to reproduce. So I just manually simulated a scenario with the same result. 

1. build a cluster with 3 fe(fe1, fe2, fe3).
2. cp fe1 and fe2 's meta dir to a new dir `meta.r` when fe1 and fe2 is running.
3. wait for 2 minutes.
4. stop fe1, fe2, fe3.
5. start fe1, fe2 with meta.r
6. after fe1 and fe2 started successfully, start fe3.

### Expected behavior (Required)
fe3 should be started successfully.

### Real behavior (Required)
```java
2021-09-26 17:13:04,734 ERROR (replayer|80) [BDBEnvironment.getDatabaseNames():346] bdb environment failure exception.
com.sleepycat.je.rep.RollbackException: (JE 7.3.7) Environment must be closed, caused by: com.sleepycat.je.rep.RollbackException: Environment invalid because
of previous exception: (JE 7.3.7) 172.26.92.154_29917_1632647567330(523):/home/disk1/gengjun/fe/meta/bdb Node 172.26.92.154_29917_1632647567330(523):/home/dis
k1/gengjun/fe/meta/bdb must rollback 1 total commits(1 of which were durable) to the earliest point indicated by transaction id=-1235545 time=2021-09-26 17:13
:00.145 vlsn=2,267,732 lsn=0x418/0x16d durable=false in order to rejoin the replication group. All existing ReplicatedEnvironment handles must be closed and r
einstantiated.  Log files were truncated to file 0x1047, offset 0x481, vlsn 2,267,730 HARD_RECOVERY: Rolled back past transaction commit or abort. Must run re
covery by re-opening Environment handles Environment is invalid and must be closed. Originally thrown by HA thread: REPLICA 172.26.92.154_29917_1632647567330(
523) Originally thrown by HA thread: REPLICA 172.26.92.154_29917_1632647567330(523) Environment invalid because of previous exception: (JE 7.3.7) 172.26.92.15
4_29917_1632647567330(523):/home/disk1/gengjun/fe/meta/bdb Node 172.26.92.154_29917_1632647567330(523):/home/disk1/gengjun/fe/meta/bdb must rollback 1 total c
ommits(1 of which were durable) to the earliest point indicated by transaction id=-1235545 time=2021-09-26 17:13:00.145 vlsn=2,267,732 lsn=0x418/0x16d durable
=false in order to rejoin the replication group. All existing ReplicatedEnvironment handles must be closed and reinstantiated.  Log files were truncated to fi
le 0x1047, offset 0x481, vlsn 2,267,730 HARD_RECOVERY: Rolled back past transaction commit or abort. Must run recovery by re-opening Environment handles Envir
onment is invalid and must be closed. Originally thrown by HA thread: REPLICA 172.26.92.154_29917_1632647567330(523) Originally thrown by HA thread: REPLICA 1
72.26.92.154_29917_1632647567330(523)
        at com.sleepycat.je.rep.RollbackException.wrapSelf(RollbackException.java:146) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.RollbackException.wrapSelf(RollbackException.java:62) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.EnvironmentImpl.checkIfInvalid(EnvironmentImpl.java:1766) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.dbi.EnvironmentImpl.checkOpen(EnvironmentImpl.java:1775) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Environment.checkOpen(Environment.java:2473) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.Environment.getDatabaseNames(Environment.java:2245) ~[je-7.3.7.jar:7.3.7]
        at com.starrocks.journal.bdbje.BDBEnvironment.getDatabaseNames(BDBEnvironment.java:339) [starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJEJournal.getMaxJournalId(BDBJEJournal.java:211) [starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.getMaxJournalId(EditLog.java:99) [starrocks-fe.jar:?]
        at com.starrocks.catalog.Catalog.getMaxJournalId(Catalog.java:5018) [starrocks-fe.jar:?]
        at com.starrocks.catalog.Catalog.replayJournal(Catalog.java:2427) [starrocks-fe.jar:?]
        at com.starrocks.catalog.Catalog$3.runOneCycle(Catalog.java:2232) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:119) [starrocks-fe.jar:?]
Caused by: com.sleepycat.je.rep.RollbackException: Environment invalid because of previous exception: (JE 7.3.7) 172.26.92.154_29917_1632647567330(523):/home/disk1/gengjun/fe/meta/bdb Node 172.26.92.154_29917_1632647567330(523):/home/disk1/gengjun/fe/meta/bdb must rollback 1 total commits(1 of which were durable) to the earliest point indicated by transaction id=-1235545 time=2021-09-26 17:13:00.145 vlsn=2,267,732 lsn=0x418/0x16d durable=false in order to rejoin the replication group. All existing ReplicatedEnvironment handles must be closed and reinstantiated.  Log files were truncated to file 0x1047, offset 0x481, vlsn 2,267,730 HARD_RECOVERY: Rolled back past transaction commit or abort. Must run recovery by re-opening Environment handles Environment is invalid and must be closed. Originally thrown by HA thread: REPLICA 172.26.92.154_29917_1632647567330(523) Originally thrown by HA thread: REPLICA 172.26.92.154_29917_1632647567330(523)
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.setupHardRecovery(ReplicaFeederSyncup.java:680) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.verifyRollback(ReplicaFeederSyncup.java:372) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.stream.ReplicaFeederSyncup.execute(ReplicaFeederSyncup.java:157) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.Replica.initReplicaLoop(Replica.java:711) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.Replica.runReplicaLoopInternal(Replica.java:474) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.Replica.runReplicaLoop(Replica.java:409) ~[je-7.3.7.jar:7.3.7]
        at com.sleepycat.je.rep.impl.node.RepNode.run(RepNode.java:1873) ~[je-7.3.7.jar:7.3.7]
```

### StarRocks version (Required)
 - all

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -210,14 +210,7 @@ public class BDBEnvironment {
                 break;
             } catch (InsufficientLogException insufficientLogEx) {
                 LOG.warn("insufficient exception, refresh and setup again", insufficientLogEx);
-                NetworkRestore restore = new NetworkRestore();
-                NetworkRestoreConfig config = new NetworkRestoreConfig();
-                config.setRetainLogFiles(false); // delete obsolete log files.
-                // Use the members returned by insufficientLogEx.getLogProviders()
-                // to select the desired subset of members and pass the resulting
-                // list as the argument to config.setLogProviders(), if the
-                // default selection of providers is not suitable.
-                restore.execute(insufficientLogEx, config);
+                refreshLog(insufficientLogEx);
                 close();
             } catch (RollbackException exception) {
                 LOG.warn("rollback exception, setup again", exception);
@@ -236,6 +229,17 @@ public class BDBEnvironment {
                 }
             }
         }
+    }
+
+    private void refreshLog(InsufficientLogException insufficientLogEx) {
+        NetworkRestore restore = new NetworkRestore();
+        NetworkRestoreConfig config = new NetworkRestoreConfig();
+        config.setRetainLogFiles(false); // delete obsolete log files.
+        // Use the members returned by insufficientLogEx.getLogProviders()
+        // to select the desired subset of members and pass the resulting
+        // list as the argument to config.setLogProviders(), if the
+        // default selection of providers is not suitable.
+        restore.execute(insufficientLogEx, config);
     }
 
     public ReplicationGroupAdmin getReplicationGroupAdmin() {
@@ -367,10 +371,7 @@ public class BDBEnvironment {
                 break;
             } catch (InsufficientLogException e) {
                 LOG.warn("catch insufficient log exception. refresh and setup again.", e);
-                NetworkRestore restore = new NetworkRestore();
-                NetworkRestoreConfig config = new NetworkRestoreConfig();
-                config.setRetainLogFiles(false);
-                restore.execute(e, config);
+                refreshLog(e);
                 close();
                 setup();
             } catch (RollbackException exception) {

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -209,10 +209,9 @@ public class BDBEnvironment {
 
                 // open epochDB. the first parameter null means auto-commit
                 epochDB = new CloseSafeDatabase(replicatedEnvironment.openDatabase(null, "epochDB", dbConfig));
-                LOG.info("everything is ok, break");
                 break;
             } catch (InsufficientLogException insufficientLogEx) {
-                LOG.warn("insufficient exception", insufficientLogEx);
+                LOG.warn("insufficient exception, refresh and setup again", insufficientLogEx);
                 NetworkRestore restore = new NetworkRestore();
                 NetworkRestoreConfig config = new NetworkRestoreConfig();
                 config.setRetainLogFiles(false); // delete obsolete log files.
@@ -223,7 +222,7 @@ public class BDBEnvironment {
                 restore.execute(insufficientLogEx, config);
                 close();
             } catch (RollbackException exception) {
-                LOG.warn("rollback exception", exception);
+                LOG.warn("rollback exception, setup again", exception);
                 close();
             } catch (DatabaseException e) {
                 LOG.warn("database exception", e);
@@ -369,7 +368,7 @@ public class BDBEnvironment {
                 names = replicatedEnvironment.getDatabaseNames();
                 break;
             } catch (InsufficientLogException e) {
-                LOG.warn("catch insufficient log exception. will recover and try again.", e);
+                LOG.warn("catch insufficient log exception. refresh and setup again.", e);
                 NetworkRestore restore = new NetworkRestore();
                 NetworkRestoreConfig config = new NetworkRestoreConfig();
                 config.setRetainLogFiles(false);
@@ -377,7 +376,7 @@ public class BDBEnvironment {
                 close();
                 setup();
             } catch (RollbackException exception) {
-                LOG.warn("rollback exception", exception);
+                LOG.warn("rollback exception, setup again", exception);
                 close();
                 setup();
             } catch (EnvironmentFailureException e) {

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -423,7 +423,7 @@ public class BDBEnvironment {
                 try {
                     db.close();
                 } catch (DatabaseException exception) {
-                    LOG.error("Error closing db", exception);
+                    LOG.error("Error closing db {}", db.getDbName(), exception);
                     closeSuccess = false;
                 }
             }
@@ -435,7 +435,7 @@ public class BDBEnvironment {
                 try {
                     epochDB.close();
                 } catch (DatabaseException exception) {
-                    LOG.error("Error closing db ", exception);
+                    LOG.error("Error closing db {}", epochDB.getDbName(), exception);
                     closeSuccess = false;
                 }
             }

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -121,62 +121,60 @@ public class BDBEnvironment {
             LOG.info("group has been reset.");
         }
 
+        // set replication config
+        replicationConfig = new ReplicationConfig();
+        replicationConfig.setNodeName(selfNodeName);
+        replicationConfig.setNodeHostPort(selfNodeHostPort);
+        replicationConfig.setHelperHosts(helperHostPort);
+        replicationConfig.setGroupName(STARROCKS_JOURNAL_GROUP);
+        replicationConfig.setConfigParam(ReplicationConfig.ENV_UNKNOWN_STATE_TIMEOUT, "10");
+        replicationConfig.setMaxClockDelta(Config.max_bdbje_clock_delta_ms, TimeUnit.MILLISECONDS);
+        replicationConfig.setConfigParam(ReplicationConfig.TXN_ROLLBACK_LIMIT,
+                String.valueOf(Config.txn_rollback_limit));
+        replicationConfig
+                .setConfigParam(ReplicationConfig.REPLICA_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
+        replicationConfig
+                .setConfigParam(ReplicationConfig.FEEDER_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
+
+        if (isElectable) {
+            replicationConfig.setReplicaAckTimeout(Config.bdbje_replica_ack_timeout_second, TimeUnit.SECONDS);
+            replicationConfig.setConfigParam(ReplicationConfig.REPLICA_MAX_GROUP_COMMIT, "0");
+            replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
+        } else {
+            replicationConfig.setNodeType(NodeType.SECONDARY);
+            replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
+        }
+
+        java.util.logging.Logger parent = java.util.logging.Logger.getLogger("com.sleepycat.je");
+        parent.setLevel(Level.parse(Config.bdbje_log_level));
+
+        // set environment config
+        environmentConfig = new EnvironmentConfig();
+        environmentConfig.setTransactional(true);
+        environmentConfig.setAllowCreate(true);
+        environmentConfig.setCachePercent(MEMORY_CACHE_PERCENT);
+        environmentConfig.setLockTimeout(Config.bdbje_lock_timeout_second, TimeUnit.SECONDS);
+        environmentConfig.setConfigParam(EnvironmentConfig.FILE_LOGGING_LEVEL, Config.bdbje_log_level);
+        if (isElectable) {
+            Durability durability = new Durability(getSyncPolicy(Config.master_sync_policy),
+                    getSyncPolicy(Config.replica_sync_policy), getAckPolicy(Config.replica_ack_policy));
+            environmentConfig.setDurability(durability);
+        }
+
+        // set database config
+        dbConfig = new DatabaseConfig();
+        dbConfig.setTransactional(true);
+        if (isElectable) {
+            dbConfig.setAllowCreate(true);
+            dbConfig.setReadOnly(false);
+        } else {
+            dbConfig.setAllowCreate(false);
+            dbConfig.setReadOnly(true);
+        }
+
         // open environment and epochDB
-        long ts = System.currentTimeMillis();
         for (int i = 0; i < RETRY_TIME; i++) {
             try {
-                LOG.info("ts: {}, iterator: {}", ts, i);
-                // set replication config
-                replicationConfig = new ReplicationConfig();
-                replicationConfig.setNodeName(selfNodeName);
-                replicationConfig.setNodeHostPort(selfNodeHostPort);
-                replicationConfig.setHelperHosts(helperHostPort);
-                replicationConfig.setGroupName(STARROCKS_JOURNAL_GROUP);
-                replicationConfig.setConfigParam(ReplicationConfig.ENV_UNKNOWN_STATE_TIMEOUT, "10");
-                replicationConfig.setMaxClockDelta(Config.max_bdbje_clock_delta_ms, TimeUnit.MILLISECONDS);
-                replicationConfig.setConfigParam(ReplicationConfig.TXN_ROLLBACK_LIMIT,
-                        String.valueOf(Config.txn_rollback_limit));
-                replicationConfig
-                        .setConfigParam(ReplicationConfig.REPLICA_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
-                replicationConfig
-                        .setConfigParam(ReplicationConfig.FEEDER_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
-
-                if (isElectable) {
-                    replicationConfig.setReplicaAckTimeout(Config.bdbje_replica_ack_timeout_second, TimeUnit.SECONDS);
-                    replicationConfig.setConfigParam(ReplicationConfig.REPLICA_MAX_GROUP_COMMIT, "0");
-                    replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
-                } else {
-                    replicationConfig.setNodeType(NodeType.SECONDARY);
-                    replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
-                }
-
-                java.util.logging.Logger parent = java.util.logging.Logger.getLogger("com.sleepycat.je");
-                parent.setLevel(Level.parse(Config.bdbje_log_level));
-
-                // set environment config
-                environmentConfig = new EnvironmentConfig();
-                environmentConfig.setTransactional(true);
-                environmentConfig.setAllowCreate(true);
-                environmentConfig.setCachePercent(MEMORY_CACHE_PERCENT);
-                environmentConfig.setLockTimeout(Config.bdbje_lock_timeout_second, TimeUnit.SECONDS);
-                environmentConfig.setConfigParam(EnvironmentConfig.FILE_LOGGING_LEVEL, Config.bdbje_log_level);
-                if (isElectable) {
-                    Durability durability = new Durability(getSyncPolicy(Config.master_sync_policy),
-                            getSyncPolicy(Config.replica_sync_policy), getAckPolicy(Config.replica_ack_policy));
-                    environmentConfig.setDurability(durability);
-                }
-
-                // set database config
-                dbConfig = new DatabaseConfig();
-                dbConfig.setTransactional(true);
-                if (isElectable) {
-                    dbConfig.setAllowCreate(true);
-                    dbConfig.setReadOnly(false);
-                } else {
-                    dbConfig.setAllowCreate(false);
-                    dbConfig.setReadOnly(true);
-                }
-
                 // open the environment
                 replicatedEnvironment = new ReplicatedEnvironment(envHome, replicationConfig, environmentConfig);
 

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -37,6 +37,7 @@ import com.sleepycat.je.rep.NoConsistencyRequiredPolicy;
 import com.sleepycat.je.rep.NodeType;
 import com.sleepycat.je.rep.ReplicatedEnvironment;
 import com.sleepycat.je.rep.ReplicationConfig;
+import com.sleepycat.je.rep.RollbackException;
 import com.sleepycat.je.rep.StateChangeListener;
 import com.sleepycat.je.rep.util.DbResetRepGroup;
 import com.sleepycat.je.rep.util.ReplicationGroupAdmin;
@@ -83,16 +84,28 @@ public class BDBEnvironment {
     private List<CloseSafeDatabase> openedDatabases;
 
     // mark whether environment is closing, if true, all calling to environment will fail
-    private boolean closing = false;
+    private volatile boolean closing = false;
 
-    public BDBEnvironment() {
+    private final File envHome;
+    private final String selfNodeName;
+    private final String selfNodeHostPort;
+    private final String helperHostPort;
+    private final boolean isElectable;
+
+
+    public BDBEnvironment(File envHome, String selfNodeName, String selfNodeHostPort,
+                          String helperHostPort, boolean isElectable) {
+        this.envHome = envHome;
+        this.selfNodeName = selfNodeName;
+        this.selfNodeHostPort = selfNodeHostPort;
+        this.helperHostPort = helperHostPort;
+        this.isElectable = isElectable;
         openedDatabases = new ArrayList<>();
         this.lock = new ReentrantReadWriteLock(true);
     }
 
     // The setup() method opens the environment and database
-    public void setup(File envHome, String selfNodeName, String selfNodeHostPort,
-                      String helperHostPort, boolean isElectable) {
+    public void setup() {
 
         this.closing = false;
 
@@ -108,60 +121,62 @@ public class BDBEnvironment {
             LOG.info("group has been reset.");
         }
 
-        // set replication config
-        replicationConfig = new ReplicationConfig();
-        replicationConfig.setNodeName(selfNodeName);
-        replicationConfig.setNodeHostPort(selfNodeHostPort);
-        replicationConfig.setHelperHosts(helperHostPort);
-        replicationConfig.setGroupName(STARROCKS_JOURNAL_GROUP);
-        replicationConfig.setConfigParam(ReplicationConfig.ENV_UNKNOWN_STATE_TIMEOUT, "10");
-        replicationConfig.setMaxClockDelta(Config.max_bdbje_clock_delta_ms, TimeUnit.MILLISECONDS);
-        replicationConfig.setConfigParam(ReplicationConfig.TXN_ROLLBACK_LIMIT,
-                String.valueOf(Config.txn_rollback_limit));
-        replicationConfig
-                .setConfigParam(ReplicationConfig.REPLICA_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
-        replicationConfig
-                .setConfigParam(ReplicationConfig.FEEDER_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
-
-        if (isElectable) {
-            replicationConfig.setReplicaAckTimeout(Config.bdbje_replica_ack_timeout_second, TimeUnit.SECONDS);
-            replicationConfig.setConfigParam(ReplicationConfig.REPLICA_MAX_GROUP_COMMIT, "0");
-            replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
-        } else {
-            replicationConfig.setNodeType(NodeType.SECONDARY);
-            replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
-        }
-
-        java.util.logging.Logger parent = java.util.logging.Logger.getLogger("com.sleepycat.je");
-        parent.setLevel(Level.parse(Config.bdbje_log_level));
-
-        // set environment config
-        environmentConfig = new EnvironmentConfig();
-        environmentConfig.setTransactional(true);
-        environmentConfig.setAllowCreate(true);
-        environmentConfig.setCachePercent(MEMORY_CACHE_PERCENT);
-        environmentConfig.setLockTimeout(Config.bdbje_lock_timeout_second, TimeUnit.SECONDS);
-        environmentConfig.setConfigParam(EnvironmentConfig.FILE_LOGGING_LEVEL, Config.bdbje_log_level);
-        if (isElectable) {
-            Durability durability = new Durability(getSyncPolicy(Config.master_sync_policy),
-                    getSyncPolicy(Config.replica_sync_policy), getAckPolicy(Config.replica_ack_policy));
-            environmentConfig.setDurability(durability);
-        }
-
-        // set database config
-        dbConfig = new DatabaseConfig();
-        dbConfig.setTransactional(true);
-        if (isElectable) {
-            dbConfig.setAllowCreate(true);
-            dbConfig.setReadOnly(false);
-        } else {
-            dbConfig.setAllowCreate(false);
-            dbConfig.setReadOnly(true);
-        }
-
         // open environment and epochDB
+        long ts = System.currentTimeMillis();
         for (int i = 0; i < RETRY_TIME; i++) {
             try {
+                LOG.info("ts: {}, iterator: {}", ts, i);
+                // set replication config
+                replicationConfig = new ReplicationConfig();
+                replicationConfig.setNodeName(selfNodeName);
+                replicationConfig.setNodeHostPort(selfNodeHostPort);
+                replicationConfig.setHelperHosts(helperHostPort);
+                replicationConfig.setGroupName(STARROCKS_JOURNAL_GROUP);
+                replicationConfig.setConfigParam(ReplicationConfig.ENV_UNKNOWN_STATE_TIMEOUT, "10");
+                replicationConfig.setMaxClockDelta(Config.max_bdbje_clock_delta_ms, TimeUnit.MILLISECONDS);
+                replicationConfig.setConfigParam(ReplicationConfig.TXN_ROLLBACK_LIMIT,
+                        String.valueOf(Config.txn_rollback_limit));
+                replicationConfig
+                        .setConfigParam(ReplicationConfig.REPLICA_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
+                replicationConfig
+                        .setConfigParam(ReplicationConfig.FEEDER_TIMEOUT, Config.bdbje_heartbeat_timeout_second + " s");
+
+                if (isElectable) {
+                    replicationConfig.setReplicaAckTimeout(Config.bdbje_replica_ack_timeout_second, TimeUnit.SECONDS);
+                    replicationConfig.setConfigParam(ReplicationConfig.REPLICA_MAX_GROUP_COMMIT, "0");
+                    replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
+                } else {
+                    replicationConfig.setNodeType(NodeType.SECONDARY);
+                    replicationConfig.setConsistencyPolicy(new NoConsistencyRequiredPolicy());
+                }
+
+                java.util.logging.Logger parent = java.util.logging.Logger.getLogger("com.sleepycat.je");
+                parent.setLevel(Level.parse(Config.bdbje_log_level));
+
+                // set environment config
+                environmentConfig = new EnvironmentConfig();
+                environmentConfig.setTransactional(true);
+                environmentConfig.setAllowCreate(true);
+                environmentConfig.setCachePercent(MEMORY_CACHE_PERCENT);
+                environmentConfig.setLockTimeout(Config.bdbje_lock_timeout_second, TimeUnit.SECONDS);
+                environmentConfig.setConfigParam(EnvironmentConfig.FILE_LOGGING_LEVEL, Config.bdbje_log_level);
+                if (isElectable) {
+                    Durability durability = new Durability(getSyncPolicy(Config.master_sync_policy),
+                            getSyncPolicy(Config.replica_sync_policy), getAckPolicy(Config.replica_ack_policy));
+                    environmentConfig.setDurability(durability);
+                }
+
+                // set database config
+                dbConfig = new DatabaseConfig();
+                dbConfig.setTransactional(true);
+                if (isElectable) {
+                    dbConfig.setAllowCreate(true);
+                    dbConfig.setReadOnly(false);
+                } else {
+                    dbConfig.setAllowCreate(false);
+                    dbConfig.setReadOnly(true);
+                }
+
                 // open the environment
                 replicatedEnvironment = new ReplicatedEnvironment(envHome, replicationConfig, environmentConfig);
 
@@ -194,8 +209,10 @@ public class BDBEnvironment {
 
                 // open epochDB. the first parameter null means auto-commit
                 epochDB = new CloseSafeDatabase(replicatedEnvironment.openDatabase(null, "epochDB", dbConfig));
+                LOG.info("everything is ok, break");
                 break;
             } catch (InsufficientLogException insufficientLogEx) {
+                LOG.warn("insufficient exception", insufficientLogEx);
                 NetworkRestore restore = new NetworkRestore();
                 NetworkRestoreConfig config = new NetworkRestoreConfig();
                 config.setRetainLogFiles(false); // delete obsolete log files.
@@ -204,7 +221,12 @@ public class BDBEnvironment {
                 // list as the argument to config.setLogProviders(), if the
                 // default selection of providers is not suitable.
                 restore.execute(insufficientLogEx, config);
+                close();
+            } catch (RollbackException exception) {
+                LOG.warn("rollback exception", exception);
+                close();
             } catch (DatabaseException e) {
+                LOG.warn("database exception", e);
                 if (i < RETRY_TIME - 1) {
                     try {
                         Thread.sleep(5 * 1000);
@@ -347,7 +369,17 @@ public class BDBEnvironment {
                 names = replicatedEnvironment.getDatabaseNames();
                 break;
             } catch (InsufficientLogException e) {
-                throw e;
+                LOG.warn("catch insufficient log exception. will recover and try again.", e);
+                NetworkRestore restore = new NetworkRestore();
+                NetworkRestoreConfig config = new NetworkRestoreConfig();
+                config.setRetainLogFiles(false);
+                restore.execute(e, config);
+                close();
+                setup();
+            } catch (RollbackException exception) {
+                LOG.warn("rollback exception", exception);
+                close();
+                setup();
             } catch (EnvironmentFailureException e) {
                 tried++;
                 if (tried == RETRY_TIME) {
@@ -394,7 +426,7 @@ public class BDBEnvironment {
                 try {
                     db.close();
                 } catch (DatabaseException exception) {
-                    LOG.error("Error closing db {}", db.getDb().getDatabaseName(), exception);
+                    LOG.error("Error closing db", exception);
                     closeSuccess = false;
                 }
             }
@@ -406,7 +438,7 @@ public class BDBEnvironment {
                 try {
                     epochDB.close();
                 } catch (DatabaseException exception) {
-                    LOG.error("Error closing db {}", epochDB.getDb().getDatabaseName(), exception);
+                    LOG.error("Error closing db ", exception);
                     closeSuccess = false;
                 }
             }
@@ -424,6 +456,7 @@ public class BDBEnvironment {
             }
             LOG.info("close replicated environment end");
         } finally {
+            closing = false;
             lock.writeLock().unlock();
         }
         return closeSuccess;

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -92,7 +92,6 @@ public class BDBEnvironment {
     private final String helperHostPort;
     private final boolean isElectable;
 
-
     public BDBEnvironment(File envHome, String selfNodeName, String selfNodeHostPort,
                           String helperHostPort, boolean isElectable) {
         this.envHome = envHome;

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBEnvironment.java
@@ -423,7 +423,7 @@ public class BDBEnvironment {
                 try {
                     db.close();
                 } catch (DatabaseException exception) {
-                    LOG.error("Error closing db {}", db.getDbName(), exception);
+                    LOG.error("Error closing db {}", db.getDatabaseName(), exception);
                     closeSuccess = false;
                 }
             }
@@ -435,7 +435,7 @@ public class BDBEnvironment {
                 try {
                     epochDB.close();
                 } catch (DatabaseException exception) {
-                    LOG.error("Error closing db {}", epochDB.getDbName(), exception);
+                    LOG.error("Error closing db {}", epochDB.getDatabaseName(), exception);
                     closeSuccess = false;
                 }
             }

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -26,9 +26,6 @@ import com.sleepycat.je.Database;
 import com.sleepycat.je.DatabaseEntry;
 import com.sleepycat.je.DatabaseException;
 import com.sleepycat.je.OperationStatus;
-import com.sleepycat.je.rep.InsufficientLogException;
-import com.sleepycat.je.rep.NetworkRestore;
-import com.sleepycat.je.rep.NetworkRestoreConfig;
 import com.starrocks.catalog.Catalog;
 import com.starrocks.common.Pair;
 import com.starrocks.common.io.DataOutputBuffer;
@@ -278,46 +275,28 @@ public class BDBJEJournal implements Journal {
         }
 
         // Open a new journal database or get last existing one as current journal database
-        Pair<String, Integer> helperNode = Catalog.getCurrentCatalog().getHelperNode();
-        List<Long> dbNames = null;
-        for (int i = 0; i < RETRY_TIME; i++) {
-            try {
-                dbNames = bdbEnvironment.getDatabaseNames();
-
-                if (dbNames == null) {
-                    LOG.error("fail to get dbNames while open bdbje journal. will exit");
-                    System.exit(-1);
-                }
-                if (dbNames.size() == 0) {
-                    /*
-                     *  This is the very first time to open. Usually, we will open a new database named "1".
-                     *  But when we start cluster with an image file copied from other cluster,
-                     *  here we should open database with name image max journal id + 1.
-                     *  (default Catalog.getCurrentCatalog().getReplayedJournalId() is 0)
-                     */
-                    String dbName = Long.toString(Catalog.getCurrentCatalog().getReplayedJournalId() + 1);
-                    LOG.info("the very first time to open bdb, dbname is {}", dbName);
-                    currentJournalDB = bdbEnvironment.openDatabase(dbName);
-                } else {
-                    // get last database as current journal database
-                    currentJournalDB = bdbEnvironment.openDatabase(dbNames.get(dbNames.size() - 1).toString());
-                }
-
-                // set next journal id
-                nextJournalId.set(getMaxJournalId() + 1);
-
-                break;
-            } catch (InsufficientLogException insufficientLogEx) {
-                // Copy the missing log files from a member of the replication group who owns the files
-                LOG.warn("catch insufficient log exception. will recover and try again.", insufficientLogEx);
-                NetworkRestore restore = new NetworkRestore();
-                NetworkRestoreConfig config = new NetworkRestoreConfig();
-                config.setRetainLogFiles(false);
-                restore.execute(insufficientLogEx, config);
-                bdbEnvironment.close();
-                bdbEnvironment.setup();
-            }
+        List<Long> dbNames = bdbEnvironment.getDatabaseNames();
+        if (dbNames == null) {
+            LOG.error("fail to get dbNames while open bdbje journal. will exit");
+            System.exit(-1);
         }
+        if (dbNames.size() == 0) {
+            /*
+             *  This is the very first time to open. Usually, we will open a new database named "1".
+             *  But when we start cluster with an image file copied from other cluster,
+             *  here we should open database with name image max journal id + 1.
+             *  (default Catalog.getCurrentCatalog().getReplayedJournalId() is 0)
+             */
+            String dbName = Long.toString(Catalog.getCurrentCatalog().getReplayedJournalId() + 1);
+            LOG.info("the very first time to open bdb, dbname is {}", dbName);
+            currentJournalDB = bdbEnvironment.openDatabase(dbName);
+        } else {
+            // get last database as current journal database
+            currentJournalDB = bdbEnvironment.openDatabase(dbNames.get(dbNames.size() - 1).toString());
+        }
+
+        // set next journal id
+        nextJournalId.set(getMaxJournalId() + 1);
     }
 
     @Override

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/BDBJEJournal.java
@@ -265,12 +265,12 @@ public class BDBJEJournal implements Journal {
     public synchronized void open() {
         if (bdbEnvironment == null) {
             File dbEnv = new File(environmentPath);
-            bdbEnvironment = new BDBEnvironment();
             Pair<String, Integer> helperNode = Catalog.getCurrentCatalog().getHelperNode();
             String helperHostPort = helperNode.first + ":" + helperNode.second;
+            bdbEnvironment = new BDBEnvironment(dbEnv, selfNodeName, selfNodeHostPort,
+                    helperHostPort, Catalog.getCurrentCatalog().isElectable());
             try {
-                bdbEnvironment.setup(dbEnv, selfNodeName, selfNodeHostPort,
-                        helperHostPort, Catalog.getCurrentCatalog().isElectable());
+                bdbEnvironment.setup();
             } catch (Exception e) {
                 LOG.error("catch an exception when setup bdb environment. will exit.", e);
                 System.exit(-1);
@@ -314,16 +314,8 @@ public class BDBJEJournal implements Journal {
                 NetworkRestoreConfig config = new NetworkRestoreConfig();
                 config.setRetainLogFiles(false);
                 restore.execute(insufficientLogEx, config);
-                if (!bdbEnvironment.close()) {
-                    LOG.error("close bdb environment failed, will exit");
-                    // NOTE: System.exit will trigger BDBEnvironment.close(),
-                    // because BDBEnvironment.close() has been registered in shutdown hook,
-                    // so in this case BDBEnvironment.close() will be called twice.
-                    // But it is ok.
-                    System.exit(-1);
-                }
-                bdbEnvironment.setup(new File(environmentPath), selfNodeName, selfNodeHostPort,
-                        helperNode.first + ":" + helperNode.second, Catalog.getCurrentCatalog().isElectable());
+                bdbEnvironment.close();
+                bdbEnvironment.setup();
             }
         }
     }

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/CloseSafeDatabase.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/CloseSafeDatabase.java
@@ -63,7 +63,7 @@ public class CloseSafeDatabase {
         return db;
     }
 
-    public String getDbName() {
+    public String getDatabaseName() {
         try {
             return db.getDatabaseName();
         } catch (Throwable t) {

--- a/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/CloseSafeDatabase.java
+++ b/fe/fe-core/src/main/java/com/starrocks/journal/bdbje/CloseSafeDatabase.java
@@ -62,4 +62,12 @@ public class CloseSafeDatabase {
     public Database getDb() {
         return db;
     }
+
+    public String getDbName() {
+        try {
+            return db.getDatabaseName();
+        } catch (Throwable t) {
+            return "";
+        }
+    }
 }

